File: 0095-不同的二叉搜索树2/Solution.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

class Solution {
    public List<TreeNode> generateTrees(int n) {
        if(n <= 0){ return new ArrayList<TreeNode>(); }
        return helper(1, n);
    }

    private List<TreeNode> helper(int index1, int index2){
        List<TreeNode> res = new ArrayList<>();
        if(index2 < index1){  res.add(null); return res; }
        if(index1 == index2){ res.add(new TreeNode(index1)); return res; }

        for(int i=index1; i<=index2; i++){
            List<TreeNode> left = helper(index1, i-1);
            List<TreeNode> right = helper(i+1, index2);
            for (TreeNode treeleft : left) {
                for (TreeNode treeright : right) {
                    TreeNode root = new TreeNode(i);
                    root.left = treeleft;
                    root.right = treeright;
                    res.add(root);
                }
            }
        }
        return res;
    }
}

File: 0095-不同的二叉搜索树2/solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def generateTrees(self, n):
        if n <= 0:
            return []
        nums = list(range(1, n+1))
        ret = self.recursion(nums)
        return ret

    def recursion(self, nums):
        if not nums:
            return [None]
        ret = []
        for i, num in enumerate(nums):
            left = self.recursion(nums[:i])
            right = self.recursion(nums[i+1:])
            for left_node in left:
                for right_node in right:
                    tmp = TreeNode(num, left=left_node, right=right_node)
                    ret.append(tmp)
        return ret

ret = Solution().generateTrees(1)
for i in ret:
    print(i)

        
